File: src/hlp/api/routers/notifications.py
"""Notification router."""

import math
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from hlp.api.deps import get_current_user, get_db
from hlp.api.schemas.common import PaginatedResponse
from hlp.api.schemas.notification import NotificationRead
from hlp.models.profile import Profile
from hlp.shared import notification_service

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get(
    "",
    response_model=PaginatedResponse[NotificationRead],
)
def list_notifications(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[Profile, Depends(get_current_user)],
    page: int = Query(1, ge=1),
    size: int = Query(25, ge=1, le=200),
) -> PaginatedResponse[NotificationRead]:
    items, total = notification_service.list_all(
        db, current_user.profile_id, page, size
    )
    pages = math.ceil(total / size) if size else 0
    return PaginatedResponse[NotificationRead](
        items=[NotificationRead.model_validate(n) for n in items],
        total=total,
        page=page,
        size=size,
        pages=pages,
    )


@router.get(
    "/unread-count",
    response_model=dict,
)
def get_unread_count(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[Profile, Depends(get_current_user)],
) -> dict:
    count = notification_service.unread_count(db, current_user.profile_id)
    return {"count": count}


@router.post(
    "/{notification_id}/read",
    response_model=NotificationRead,
)
def mark_notification_read(
    notification_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[Profile, Depends(get_current_user)],
) -> NotificationRead:
    notif = notification_service.mark_read(
        db, notification_id, current_user.profile_id
    )
    db.commit()
    db.refresh(notif)
    return NotificationRead.model_validate(notif)


@router.post(
    "/read-all",
    response_model=dict,
)
def mark_all_notifications_read(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[Profile, Depends(get_current_user)],
) -> dict:
    count = notification_service.mark_all_read(db, current_user.profile_id)
    db.commit()
    return {"updated": count}

File: frontend/src/api/notifications.ts
import { api } from './client'
import type { PaginatedResponse } from './types'

export interface NotificationRead {
  notification_id: number
  profile_id: number
  pricing_request_id: number
  title: string
  message: string
  read: boolean
  created_at: string
}

export interface ListNotificationsParams {
  page?: number
  size?: number
}

export async function listNotifications(
  params: ListNotificationsParams = {},
): Promise<PaginatedResponse<NotificationRead>> {
  const { data } = await api.get<PaginatedResponse<NotificationRead>>(
    '/api/notifications',
    { params },
  )
  return data
}

export async function getUnreadCount(): Promise<{ count: number }> {
  const { data } = await api.get<{ count: number }>('/api/notifications/unread-count')
  return data
}

export async function markRead(id: number): Promise<NotificationRead> {
  const { data } = await api.post<NotificationRead>(`/api/notifications/${id}/read`)
  return data
}

export async function markAllRead(): Promise<{ updated: number }> {
  const { data } = await api.post<{ updated: number }>('/api/notifications/read-all')
  return data
}
